docs/operators/deduper.md

```diff
@@ -0,0 +1,523 @@
+===================================
+Class Structure
+----------------------------
+
+![](images/deduper/image03.png)
+
+---------------
+
+Architectural Details
+----------------------------------
+
+![](images/deduper/image02.png)
+
+Concepts
+---------------------
+
+### Dedup Key - Specified by *keyExpression* parameter
+
+A dedup key is a set of one or more fields in the data tuple which
+acts as the key for the tuples.
+This is used by the deduper to compare tuples to arrive at the
+conclusion on whether two tuples are duplicates. If Dedup Key of two
+tuples match, then they are duplicates, else they are unique.
+
+### Expiry Key - Specified by *timeExpression* parameter
+
+A tuple may or may not have an Expiry Key. Dedup operator cannot
+keep storing all the data that is flowing into the operator. At some
+point it becomes essential to discard some of the historical tuples in
+interest of memory and efficiency.
+
+At the same time, tuples are expected to arrive at the Dedup
+operator within some time after they are generated. After this time, the
+tuples may be considered as stale or obsolete.
+
+In such cases, the Deduper chooses to consider these tuples
+expired and takes no action but to
+separate out these tuples on a different port in order to be processed
+by some other operator or offline analysis.
+
+In order to create a criterion for discarding such tuples, we
+introduce an Expiry Key. Looking at the value of the Expiry Key in each
+tuple, we can decide whether or not to discard this tuple as
+expired.
+
+The expiry key that we consider in Time Based Dedup is
+time. This usually works with
+another parameter called Expiry Period defined next.
+
+### Expiry Period
+
+The Expiry Period is the value supplied by the user which decides
+when a particular tuple expires.
+
+### Time Points
+
+For every dataset that the deduper processes, it maintains a set
+of time points:
+
+1.  *Latest Point* - This is the maximum
+    time point observed in all the processed tuples.
+2.  *Expiry Point* - This is given by:
+    `Expiry Point = Latest Point - Expiry`
+    Period
+
+These points help the deduper to make decisions related to expiry
+of a tuple.
+
+### Example - Expiry
+
+|Tuple Id|Expiry Key (Expiry Period = 10)|Latest Point|Expiry Point|Decision for Tuple|
+|:------:|:-----------------------------:|:----------:|:----------:|:----------------:|
+|1|10|10|1|Not Expired|
+|2|20|20|11|Not Expired|
+|3|25|25|16|Not Expired|
+|4|40|40|31|Not Expired|
+|5|21|40|31|Expired|
+|6|35|40|31|Not Expired|
+|7|45|45|36|Not Expired|
+|8|57|57|48|Not Expired|
+
+
+### Time Buckets (A component of Managed State)
+
+One of the requirements of the Deduper is to store all the unique
+tuples (actually, just the keys of tuples). Keeping an ever growing
+cache in memory is not scalable. So what we need is a limited cache
+backed by a persistent store. Now to reduce cache misses, we load a
+chunk of data (called Buckets), together into memory. Buckets help
```

Comment:
No, entire buckets are not loaded. This was missed while adopting the documentation for managed state. Will correct it.